Fix annotating sites with broken Function.bind polyfills

seanh opened the pull request:
Fixes https://github.com/hypothesis/client/issues/245

seanh commented on src/annotator/test/annotation-sync-test.js:
```diff
@@ -45,96 +45,115 @@ describe('AnnotationSync', function() {
-        createAnnotationSync();
-
-        publish('deleteAnnotation', {msg: ann}, function() {});
+  function addTests() {
```
The problem here is that I want to run all of `AnnotationSync`'s tests in two contexts: with and without a broken `Function.bind` in the parent page. Maybe not _all_ the tests really need to be run with the broken `Function.bind`, but a few of them should be, and this avoids duplicating them.

It does add some "cleverness" to the test code though.

An alternative would be to just add _one_ test for the broken Function.bind context: that the constructor doesn't crash. Since the broken Function.bind that the tests are using is one that crashes when called, just this one test would prove that `AnnotationSync` doesn't use the document's Function.bind

robertknight commented on src/annotator/annotation-sync.coffee:
```diff
@@ -21,13 +21,18 @@ module.exports = class AnnotationSync
+    iframe = document.createElement('iframe')
+    iframe.style.display = 'none'
+    document.body.appendChild(iframe)
+    bind = iframe.contentWindow.Function.prototype.bind
```
I think this would make sense. It seems likely that we might want to do this in other places in future. It should also be much easier to test if you extract it out into a separate module.

seanh commented:
Most of the tests diff that you see is just due to code having been indented. I didn't add or change any tests. I just wrapped all the tests in an `addTests()` function (indenting them all), called `addTests()` at the bottom of the file, and then added a `context()` for when the parent page contains a broken `Function.bind` polyfill (just a `context()` containing a `beforeEach()` and an `afterEach()`) and called `addTests()` a second time from within that context.

As noted above, I think it might be better just to add a single "it does not crash if the parent page contains a crashing Function.bind polyfill" test, instead of doing this `addTests()` thing.

seanh commented:
Ok, I've moved the clean function.Bind into a clean-context util module, and I've replaced the test cleverness discussed above. Ready for review again

robertknight commented on src/annotator/util/clean-context.js:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Export clean versions of standard functions and objects.
+ *
```
I think we need to be a bit clearer about the motivation here and what "clean" means. Specifically that this module is a workaround for environments that overwrite properties on global objects with broken polyfills or other problematic values.

seanh replied:
I've edited the paragraph below to make this more explicit

robertknight commented on src/annotator/util/clean-context.js:
```diff
@@ -0,0 +1,22 @@
+ * This module exports "clean" versions of various standard functions that may
+ * have been modified in the parent browsing context that the client is
+ * injected into. Code should always use the clean versions from this module,
+ * and never the potentially modified versions from the parent context.
```
Need to clarify what "Code" means here. Specifically you mean code in the "annotator/" dir, as this isn't relevant for the sidebar.

seanh replied:
Done, although I don't want to mention the annotator dir by name here as that comment seems unlikely to be updated if we rename it

robertknight commented on src/annotator/util/clean-context.js:
```diff
@@ -0,0 +1,22 @@
+// browsing context to extract clean objects and functions from.
+var iframe = document.createElement('iframe');
+iframe.style.display = 'none';
+document.body.appendChild(iframe);
```
On my shiny new Mac I measured the cost of creating an iframe as being ~6-13ms, so probably up ~100ms on a mobile device. It would be nice not to do this for the vast majority of pages that are not going to need it.

I've asked around whether there is a cheaper way, but an approach I'm pretty sure will work fine for feature detection is just to check whether the stringified function (via `fn.toString()`) contains the text `[native code]`. This is currently a convention but [will most likely be part of the language spec](https://tc39.github.io/Function-prototype-toString-revision/) in future.

seanh replied:
That looks like it's `"NativeFunction"` rather than `"[native code]"`? But is there any guarantee that a page's replacement of a native function won't stringify to `"NativeFunction"` as well?

robertknight replied:
@seanh "NativeFunction" in the spec is a reference to another part of the spec. Search for other occurrences of "NativeFunction" in that document :)

robertknight replied:
Yes it is technically possible for a page to overwrite `toString()` with something which always returns `function () { [native code] }` but I think it is unlikely enough that we'll be OK doing that.

seanh replied:
Or of someone just modifies `Function.prototype.bind.something` but doesn't change `toString`?

Let's see if we run into anyone that does it, I guess. I think this clean context thing is a temporary batch (before moving all of `src/annotator` into a same-origin iframe) anyway.

If we want to move the whole thing into an iframe by the way, when we do that we will have to eat this 6-100ms delay every time anyway won't we?

seanh replied:
Done - I've changed it to not inject an iframe if it doesn't think bind has been modified

robertknight commented:
I tested this by overwriting `Function.prototype.bind` with a function that just throws an exception and found a couple more places that blow up `_setupSidebarEvents` and `_setupDocumentEvents`.

seanh commented:
> I tested this by overwriting Function.prototype.bind with a function that just throws an exception and found a couple more places that blow up _setupSidebarEvents and _setupDocumentEvents.

Fixed these two as well

seanh commented on src/annotator/test/annotation-sync-test.js:
```diff
@@ -136,5 +136,24 @@ describe('AnnotationSync', function() {
       });
     });
+
+    context('if the parent context has a crashing Function.bind polyfill', function() {
```
I've left this test (and the one for sidebar below) here as I can't think of a better way. This obviously isn't ideal as every module that uses clean-context would need to have its own "doesn't crash" test.

I looked into replace `Function.prototype.bind` globally with a crashing function in the tests, but don't see any way to do that for all non-sidebar tests but _not_ do it for the `src/sidebar` tests.

I looked into writing an eslint rule to warn on uses of bind, but it seemed difficult to warn when `bind()` or `bind.call()` is called but _not_ when the calling code is in `src/sidebar` and _not_ when `cleanContext.bind.call()` is called. (It may be doable, I've not written eslint rules before, but there wasn't an obvious good way to do it.)

The remaining option would be to delete these tests, but that would simply mean that the actual bug fixes in this pull request would not be covered by tests. I don't think the presence of these tests is doing any particular harm or means that we'd necessarily have to add similar tests for every other piece of code that uses clean-context.

seanh commented:
@robertknight This can be reviewed again.

My one remaining concern is whether CSP can prevent the iframe from being injected? I tested in Chrome using the Chrome extension on a page with `<meta http-equiv="Content-Security-Policy" content="child-src 'none';">` and it works fine. Appending a same-origin iframe seems to be allowed.

The bookmarklet fails on that page (either Firefox or Chrome) but that's not this branch's fault - app.html runs afoul of the page's content security policy,

If I set an `src="http://www.example.com"` on the iframe (to simulate what would happen if Chrome did block it) then on the page with the CSP, using the Chrome extension, it gets a CSP error that entirely prevents the sidebar from loading.

What I'd like to do is catch this CSP `DOMException` and just fall back on using the parent context's bind, and a normal `try ... catch` doesn't seem to catch this exception.

seanh commented:
(My theory by the way is that since the iframe has _no_ `src` at all is always allowed. If I set `iframe.src` to `'http://localhost:8889/'` (which is there I'm serving the page with the CSP) that is blocked. And if I change the CSP from none to self it's not blocked anymore. But as long as there _no_ src at all Chrome at least seems to allow it regardless of CSP.

seanh commented:
Closed in favour of https://github.com/hypothesis/client/pull/331
